src/lib/Microsoft.Health.Fhir.Ingest.Template/CalculatedFunction/CalculatedFunctionContentTemplate.cs

```diff
@@ -0,0 +1,210 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using EnsureThat;
+using Microsoft.Health.Fhir.Ingest.Data;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Fhir.Ingest.Template.CalculatedFunction
+{
+    public class CalculatedFunctionContentTemplate : IContentTemplate
+    {
+        public const string MatchedToken = "matchedToken";
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string TypeName { get; set; }
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string TypeMatchExpression { get; set; }
+
+        public virtual ExpressionLanguage? TypeMatchExpressionLanguage { get; set; }
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string DeviceIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? DeviceIdExpressionLanguage { get; set; }
+
+        public virtual string PatientIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? PatientIdExpressionLanguage { get; set; }
+
+        public virtual string EncounterIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? EncounterIdExpressionLanguage { get; set; }
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string TimestampExpression { get; set; }
+
+        public virtual ExpressionLanguage? TimestampExpressionLanguage { get; set; }
+
+        public virtual string CorrelationIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? CorrelationIdExpressionLanguage { get; set; }
+
+#pragma warning disable CA2227
+        public virtual IList<CalculatedFunctionValueExpression> Values { get; set; }
+#pragma warning restore CA2227
+
+        [JsonConverter(typeof(StringEnumConverter))]
+        public virtual ExpressionLanguage DefaultExpressionLanguage { get; set; } = ExpressionLanguage.JsonPath;
+
+        [JsonIgnore]
+        public virtual IExpressionEvaluatorFactory ExpressionEvaluatorFactory { get; set; } = new CachingExpressionEvaluatorFactory();
+
+        public virtual IEnumerable<Measurement> GetMeasurements(JToken token)
+        {
+            EnsureArg.IsNotNull(token, nameof(token));
+            EnsureArg.IsTrue(token is JObject);
```

Comment:
Based on the [IConverter](https://github.com/microsoft/iomt-fhir/blob/7794cbcc463e8d26c3097cd5e2243d770f26fe45/src/lib/Microsoft.Health.Fhir.Ingest/Data/EventDataWithJsonBodyToJTokenConverter.cs) we are using for Normalization, it appears that IContentTemplate will always receive a `JObject` here. I enforce this as I use functions related to JObjects later on in the code.

Reply:
I believe there is an EnsureArg function to enforce a type.